FakeXrmEasy.Shared/Metadata/MetadataGenerator.cs

```diff
@@ -0,0 +1,198 @@
+                            if (peerProperty == null || peerProperty.PropertyType.Name == "IEnumerable`1") // N:N relationship
+                            {
+                                ManyToManyRelationshipMetadata relationshipMetadata = new ManyToManyRelationshipMetadata();
+                                relationshipMetadata.SchemaName = relationshipSchemaNameAttribute.SchemaName;
+
+                                manyToManyRelationshipMetadatas.Add(relationshipMetadata);
+                            }
+                            else // 1:N relationship
+                            {
+                                AddOneToManyRelationshipMetadata(earlyBoundEntity, property, property.PropertyType.GetGenericArguments()[0], peerProperty, oneToManyRelationshipMetadatas);
+                            }
+                        }
+                        else //N:1 Property
+                        {
+                            AddOneToManyRelationshipMetadata(property.PropertyType, property.PropertyType.GetProperties().SingleOrDefault(x => x.PropertyType.GetGenericArguments().SingleOrDefault() == earlyBoundEntity && GetCustomAttribute<RelationshipSchemaNameAttribute>(x)?.SchemaName == relationshipSchemaNameAttribute.SchemaName), earlyBoundEntity, property, manyToOneRelationshipMetadatas);
+                        }
+                    }
+                }
+                if (attributeMetadatas.Any())
+                {
+                    metadata.SetSealedPropertyValue("Attributes", attributeMetadatas.ToArray());
+                }
+                if (manyToManyRelationshipMetadatas.Any())
+                {
+                    metadata.SetSealedPropertyValue("ManyToManyRelationships", manyToManyRelationshipMetadatas.ToArray());
+                }
+                if (manyToOneRelationshipMetadatas.Any())
+                {
+                    metadata.SetSealedPropertyValue("ManyToOneRelationships", manyToOneRelationshipMetadatas.ToArray());
+                }
+                if (oneToManyRelationshipMetadatas.Any())
+                {
+                    metadata.SetSealedPropertyValue("OneToManyRelationships", oneToManyRelationshipMetadatas.ToArray());
+                }
+                entityMetadatas.Add(metadata);
+            }
+            return entityMetadatas;
+        }
+
+        private static T GetCustomAttribute<T>(MemberInfo member) where T : Attribute
+        {
+            return (T)Attribute.GetCustomAttribute(member, typeof(T));
+        }
+
+        private static AttributeMetadata CreateAttributeMetadata(Type propertyType)
+        {
+            switch (propertyType.Name)
+            {
+                case "String":
```

Comment:
Just some giving some feedback... and this is totally subjective but I'd rather prefer comparing using the property type as opposed to use magic strings? i.e. 

    if (attributeType == typeof(string)) {

     }

I also believe it would be nice to add one unit test per data type so we don't miss any edge case (i.e. String type in early bound to String Metadata, OptionSetValue to PicklistAttributeMetadata etc...

Thoughts?

Reply:
Hello Jordi,
I refactored the function to remove the magic strings, and I introduced some other magic strings in the process 😄 for Id, statecode and statuscode.
I will add some more unit tests for each type... my Surface goes on fire every time I open the GeneratedCode.cs file 😈 so it will take some time to find one attribute per data type I guess.